### ISSUE TITLE
code-duplication-control: normalizeCode() strips whitespace before comments, so an all-comment block collapses to an empty-string key and collides with any other file's comment-led block

### ISSUE THREAD

**Issue by ruleofcodedev** [Rule-Of-Code/RoC#126]
**Version:** 7.21.0
**Law:** `code-duplication-control` — `CodeDuplicationControlLaw.normalizeCode()` in `dist/checkers/code-quality-laws/code-duplication-control.js`

Kind: it reports code that is correct — a false positive between two files that share no real logic at all.

## What happened

```
1. Code Duplication Control [1 violations] (Score: 95/100)
   • Code duplication found in: apps/festrr/src/app/app.config.ts, libs/shell/feature/src/lib/launch/session-start.guard.ts

   💡 What the law is asking for:
   • Consider eslint-plugin-sonarjs for advanced duplication detection
   • Extract common code into shared functions or modules
   • Use inheritance or composition to reduce duplication
   ... and 1 more — run with --verbose for all, or --export json for the full result
```

`--verbose` and `--export json` both repeat the same one-line finding — no line numbers, no snippet. Diagnosing *why* these two files matched required reading `normalizeCode()`'s own source and reproducing the collision by hand; the audit output gives nothing to go on.

`app.config.ts` is a flat Angular provider array — no function, no arrow function, nothing resembling `session-start.guard.ts`'s `CanActivateFn`. `jscpd`, run separately in the same pipeline, finds zero overlap between them.

## What is wrong

```js
static normalizeCode(code) {
    return code
        .replace(/\s+/g, ' ')       // (1) collapses ALL whitespace, incl. newlines — first
        .replace(/\/\/.*$/gm, '')   // (2) strip `//` comments — but (1) already erased every
                                    //     line boundary `$` could anchor to, so this now matches
                                    //     from the FIRST `//` to the end of the whole string
        .replace(/\/\*[\s\S]*?\*\//g, '')
        .replace(/\b\d+\b/g, 'NUM')
        .replace(/['"`][^'"`]*['"`]/g, 'STR')
        .trim();
}
```

Once whitespace is collapsed first, step (2)'s `/gm` comment-strip has nothing left to stop at and eats everything after the first `//`, comments and real code alike. Any block whose first line is a `//` comment normalizes to `""`.

A second contributor in `extractCodeBlocks()`: the block-start trigger is `trimmed.includes('function') || trimmed.includes('=>')` — a plain substring test, so it fires inside comments too (e.g. `// see each function's own doc comment`). That's what made both of my files' extracted blocks open with a `//` line in the first place — a very ordinary shape for any documented function.

## Reproduction

Traced by calling the law's own exported static methods directly, not guessed:

```js
const { CodeDuplicationControlLaw } = require('ruleofcode/dist/checkers/code-quality-laws/code-duplication-control.js');
const fs = require('fs');

for (const file of ['apps/festrr/src/app/app.config.ts',
                     'libs/shell/feature/src/lib/launch/session-start.guard.ts']) {
  const content = fs.readFileSync(file, 'utf8');
  for (const block of CodeDuplicationControlLaw.extractCodeBlocks(content)) {
    console.log(file, 'raw.length=', block.length,
                'normalized=', JSON.stringify(CodeDuplicationControlLaw.normalizeCode(block)));
  }
}
```

```
apps/festrr/src/app/app.config.ts                              raw.length= 1084  normalized= ""
libs/shell/feature/src/lib/launch/session-start.guard.ts        raw.length= 429   normalized= ""
```

Both raw blocks clear the configured `minDuplicationBlockSize` (250 in this project's config), both normalize to the empty string, both land under the same key in `analyzeDuplication()`'s map, and since that key now has `files.length > 1` it's reported as duplication — of nothing.

For comparison: `npx jscpd` (this project also runs it, separately, in the same `npm run verify`) reports 5 real clones elsewhere in the repo and finds zero overlap between these two files.

## What it cost me

A real PR (`feature/fstr-406`) sat red against a false positive with no line-level detail to go on. Fixing my own code could not have helped — the two files share no logic — so the only path forward was either suppress the law for the new file with a config exception (which I did, with the root cause written into the `$comment`), or read `dist/` to prove it wasn't real. I did both, but the second one is not something every contributor filing a PR should have to do.

## Is the same logic somewhere else?

Checked: `grep -rl normalizeCode node_modules/ruleofcode/dist` matches only this one file, so this specific helper is not duplicated elsewhere in the package. I did not check whether other checkers have their *own*, differently-named whitespace/comment-normalization helper with the same ordering mistake — I only looked for this exact function name.

## Suggested fix

1. Reorder `normalizeCode()` — strip comments before collapsing whitespace:
   ```js
   static normalizeCode(code) {
       return code
           .replace(/\/\/.*$/gm, '')
           .replace(/\/\*[\s\S]*?\*\//g, '')
           .replace(/\s+/g, ' ')
           .replace(/\b\d+\b/g, 'NUM')
           .replace(/['"`][^'"`]*['"`]/g, 'STR')
           .trim();
   }
   ```
2. Belt-and-braces: skip an empty normalized key in `analyzeDuplication()` (`if (key.length === 0) continue;`) — an all-comment block was never real duplication even before the ordering bug.
3. Lower priority, separate from (1): `extractCodeBlocks()`'s substring trigger matches inside comments, which is what produces comment-led blocks in the first place. Worth a follow-up, but (1) also protects any legitimately comment-led *real* block, so it's the one that matters most.
4. The finding itself printing matched line ranges (the way `jscpd`'s own output does) rather than only a file list would have made this diagnosable from the audit output alone, in seconds rather than a standalone repro script against `dist/`.

## What I did NOT verify

- Whether this reproduces on 7.22.0 (I'm on 7.21.0, pinned in this project's `package.json`) — I have not re-staged on the latest version.
- Whether other checkers hold a differently-named copy of the same collapse-then-strip ordering mistake; I only grepped for the literal name `normalizeCode`.
- Windows-specific line-ending effects (`\r\n` vs `\n`) on `extractCodeBlocks()`'s line-splitting — this project runs on Windows (git-bash) and I have not tried the same repro on a Unix checkout to rule out a platform-specific contribution, though the whitespace-collapse ordering bug reproduces regardless.

## My config

```json
"code-quality": {
  "maxComplexity": 20,
  "minFunctionComplexity": 10,
  "maxDependencies": 50,
  "minDuplicationBlockSize": 250,
  "minCommentRatioPercent": 10,
  "minFileSizeForComments": 50
}
```

**Comment by ruleofcodedev** [comment]
Fixed in #133, merged to `develop`. Ships in the next release.

Your diagnosis was exact. With whitespace collapsed first, `$` under `/m` had nothing left to stop at, so the line-comment strip ate from the first `//` to the end of the block.

Two changes rather than one: the strip order is fixed, and a guard at the call site skips a block that normalises to less than `minBlockSize`. The threshold measures the **raw** block, so a comment-only block passed it while carrying no logic at all — without that guard the same collision returns the moment a block normalises to something short but non-empty.
